packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/mongo/MongoCollectionTests.kt

```diff
@@ -0,0 +1,1466 @@
+
+            val mongoDBClientParent = retry(
+                action = { parentCollection.findOne() },
+                until = { response -> response != null }
+            )
+            assertEquals(syncedParent._id, mongoDBClientParent!!._id)
+            assertEquals(syncedParent.any!!.asRealmObject<ChildCollectionDataType>()._id, mongoDBClientParent.any!!.asRealmObject<ChildCollectionDataType>()._id)
+        }
+    }
+
+    @Test
+    open fun findOne_typedLinks_throwsOnMissingTargetSchema() = runBlocking<Unit> {
+        Realm.open(
+            SyncConfiguration.Builder(user, FLEXIBLE_SYNC_SCHEMA)
+                .initialSubscriptions {
+                    add(it.query<ParentCollectionDataType>())
+                    add(it.query<ChildCollectionDataType>())
+                }
+                .build()
+        ).use { realm ->
+            val syncedParent = realm.write {
+                copyToRealm(
+                    ParentCollectionDataType().apply {
+                        any = RealmAny.create(ChildCollectionDataType())
+                    }
+                )
+            }
+            // We need to upload schema before proceeding
+            realm.syncSession.uploadAllLocalChanges(30.seconds)
+
+            @OptIn(ExperimentalKBsonSerializerApi::class)
+            val parentCollection = collection<ParentCollectionDataType>(
+                EJson(
+                    serializersModule = realmSerializerModule(
+                        setOf(ParentCollectionDataType::class)
+                    )
+                )
+            )
+
+            assertFailsWithMessage<SerializationException>("Cannot resolve target class in schema: Unknown class '${"$"}ref=ChildCollectionDataType'") {
+                // We need to await until the response is non-null otherwise there will be nothing
+                // to deserialize and no exception will be thrown
+                retry(
+                    action = { parentCollection.findOne() },
+                    until = { response -> response != null }
+                )
+            }
+        }
+    }
+
+    @OptIn(ExperimentalKBsonSerializerApi::class)
+    @Test
+    open fun findOne_embeddedObjects() = runBlocking<Unit> {
+        // Empty collections
+        assertNull(collection.findOne())
+
+        val parentCollection = collection<ParentCollectionDataType>()
+        parentCollection.insertOne(
+            ParentCollectionDataType().apply {
+                embeddedChild = EmbeddedChildCollectionDataType().apply { name = "EMBEDDED-NAME" }
+            }
+        )
+
+        parentCollection.find().single().run {
+            assertEquals("EMBEDDED-NAME", embeddedChild!!.name)
+        }
+    }
+
+    @OptIn(ExperimentalKBsonSerializerApi::class)
+    @Test
+    fun findOne_extraFieldsAreDiscarded() = runBlocking<Unit> {
+        val withDocumentClass = collection.withDocumentClass<BsonDocument>()
+        withDocumentClass.insertOne(
+            BsonDocument(
```

Comment:
It would be great if it could be written as:
```
Bson {
    "_id" to Random.nextInt(),
    "name" to "object-1",
    "extra" to "extra",
}
```

Reply:
I just added this convenience extension constructor in https://github.com/realm/realm-kotlin/pull/1593/files#diff-62866c3bf368e1c82bfc172e3aa985f0d097e7ae6bb93002d12e5b78978c235dR1418. More clever DSL-construct should probably be filed as an issue in https://github.com/mongodb/kbson